// strong-trace-waterfalltransform/lib/summary.js
"use strict";

// note that summary views must be called AFTER subsegments are computed
module.exports = function waterfallSummary(w, options){
  if( w.segments && w.segments.length === 0){
    return
  }

  var factor = (options && options.summaryFactor) || 1000  //for a 1000 pixel output
  var threshold = w.maxEnd / factor
  var summary = {}
  var tree          = w.tree

  summary.subSegments = []
  summary.start = 0
  summary.end = w.maxEnd

  tree.children.forEach(dft)

  w.summary = summary
  function dft(tree){
    var cost,
        s,
        sub
    if( !tree.id ){
      return
    }

    s = tree.segment
    cost = s.end - s.start
    if( tree.children.length && cost > threshold ){
      summary.subSegments.concat(s.subSegments)
      tree.children.forEach(dft)
    } else {
      sub = {
        start     : s.start,
        end       : s.end,
        exclusive : s.end - s.start,
        parent    : s,
        uid        : s.uid + ':agg'
      }
      summary.subSegments.push(sub)
    }
  }
}
