// cxviz-timeseries/lib/schema.js
'use strict';

var d3     = require('d3')
var format = require('cxviz-format')

module.exports = Schema

function Schema(options){
  if (!(this instanceof Schema)){
    return new Schema(options)
  }
  var self = this
  self.base = (options && options.keySchema) || {}
  self.color =  options.color || d3.scale.category10()

  //keep helper functions in the closure so that the d3 'this' pointer can be used
  self.y = function y(d){
    return (self.base[d.name] && self.base[d.name].y) || 'y'
  }

  self.lineClass = function lineClass(d){
    return (self.base[d.name] && self.base[d.name].class) || 'gradientLine'
  }

  self.stroke = function stroke(d){
    return self.color(d.name)
  }

  self.type = function type(d){
    return self.typeByKey(d.name)
  }

  self.typeByKey = function typeByKey(key){
    return (self.base[key] && self.base[key].type) || 'line'
  }

  self.formatByKey = function formatByKey(key){
    var y = self.y({name: key})
    if (options.format && options.format[y] && format[options.format[y]]) {
      return format[options.format[y]]
    } else {
      return format.num
    }
  }
}
